File: src/lib/algorithms/atkinson.ts
// Atkinson dithering algorithm
export function atkinsonDithering(
    grayscale: Uint8ClampedArray,
    width: number,
    height: number,
    threshold: number = 128
  ): ImageData {
    // Create output ImageData
    const outputData = new ImageData(width, height);
    const output = outputData.data;
    
    // Create a copy of the grayscale image to apply error diffusion
    const buffer = new Float32Array(grayscale);
    
    // Apply dithering
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const idx = y * width + x;
        const oldPixel = buffer[idx];
        
        // Apply threshold
        const newPixel = oldPixel < threshold ? 0 : 255;
        
        // Calculate error
        const error = (oldPixel - newPixel) / 8; // Distribute 1/8 of error to each neighbor
        
        // Distribute error to neighboring pixels using Atkinson pattern
        // As per standard: 1/8 of error to each of the 6 neighbors:
        // (x+1,y), (x+2,y), (x-1,y+1), (x,y+1), (x+1,y+1), (x,y+2)

        // Pixel (x+1, y)
        if (x + 1 < width) {
          buffer[idx + 1] += error;
        }
        
        // Pixel (x+2, y)
        if (x + 2 < width) {
          buffer[idx + 2] += error;
        }
        
        // Pixel (x-1, y+1)
        if (y + 1 < height && x - 1 >= 0) {
          buffer[idx + width - 1] += error;
        }
        
        // Pixel (x, y+1)
        if (y + 1 < height) { // x is implicitly valid within loop bounds
          buffer[idx + width] += error;
        }
        
        // Pixel (x+1, y+1)
        if (y + 1 < height && x + 1 < width) {
          buffer[idx + width + 1] += error;
        }
        
        // Pixel (x, y+2)
        if (y + 2 < height) { // x is implicitly valid within loop bounds
          buffer[idx + 2 * width] += error; 
        }
        
        // Set all RGB channels to the result (black or white)
        output[idx * 4] = newPixel;     // R
        output[idx * 4 + 1] = newPixel; // G
        output[idx * 4 + 2] = newPixel; // B
        output[idx * 4 + 3] = 255;      // A (fully opaque)
      }
    }
    
    return outputData;
  }

File: rust_wasm/src/atkinson.rs
// Atkinson dithering algorithm
pub fn atkinson_dither(
    input_ptr: *const u8,
    width: usize,
    height: usize,
    threshold: u8,
    output_ptr: *mut u8,
) {
    // Convert the input and output pointers to slices
    let input = unsafe { std::slice::from_raw_parts(input_ptr, width * height) };
    let output = unsafe { std::slice::from_raw_parts_mut(output_ptr, width * height) };
    
    // Create a buffer for error diffusion
    let mut buffer = vec![0.0f32; width * height];
    
    // Initialize buffer with input values
    for i in 0..input.len() {
        buffer[i] = input[i] as f32;
    }
    
    // Apply dithering
    for y in 0..height {
        for x in 0..width {
            let idx = y * width + x;
            let old_pixel = buffer[idx];
            
            // Apply threshold
            let new_pixel = if old_pixel < threshold as f32 { 0.0 } else { 255.0 };
            
            // Calculate error and distribute 1/8 of error to each neighbor
            let error = (old_pixel - new_pixel) / 8.0;
            
            // Distribute error to neighboring pixels using Atkinson pattern
            if x + 1 < width {
                buffer[idx + 1] += error;
            }
            
            if x + 2 < width {
                buffer[idx + 2] += error;
            }
            
            if y + 1 < height {
                if x > 0 {
                    buffer[idx + width - 1] += error;
                }
                
                buffer[idx + width] += error;
                
                if x + 1 < width {
                    buffer[idx + width + 1] += error;
                }
            }
            
            if y + 2 < height {
                buffer[idx + width * 2] += error;
            }
            
            // Set the output pixel
            output[idx] = new_pixel as u8;
        }
    }
} 